File: sol/solution/0001-0100/0021. Merge Two Sorted Lists/Solution.erl
%% Definition for singly-linked list.
%%
%% -record(list_node, {val = 0 :: integer(),
%%                     next = null :: 'null' | #list_node{}}).
-spec merge_two_lists(List1 :: #list_node{} | null, List2 :: #list_node{} | null) -> #list_node{} | null.
merge_two_lists(null, List2) -> List2;
merge_two_lists(List1, null) -> List1;
merge_two_lists(List1 = #list_node{val = V1, next = Next1}, List2 = #list_node{val = V2, next = Next2}) ->
    if
        V1 =< V2 ->
            #list_node{val = V1, next = merge_two_lists(Next1, List2)};
        true ->
            #list_node{val = V2, next = merge_two_lists(List1, Next2)}
    end.

File: sol/solution/0001-0100/0021. Merge Two Sorted Lists/Solution.ex
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end
defmodule Solution do
  @spec merge_two_lists(list1 :: ListNode.t | nil, list2 :: ListNode.t | nil) :: ListNode.t | nil
  def merge_two_lists(nil, list2), do: list2
  def merge_two_lists(list1, nil), do: list1
  def merge_two_lists(%{val: v1, next: n1} = list1, %{val: v2, next: n2} = list2) do
    if v1 <= v2 do
      %ListNode{val: v1, next: merge_two_lists(n1, list2)}
    else
      %ListNode{val: v2, next: merge_two_lists(list1, n2)}
    end
  end
end